### DISCUSSION TITLE
Dynamic Facts

### DISCUSSION THREAD

**Discussion by an-ky** [pyinfra-dev/pyinfra#1239] [Q&A] [answered]
Hi,

I have the following deployment to setup a virtual machine using `virsh`.

```
@deploy("Install virtual_machine")
def prepare_virtual_machine(
    virtual_machine: DotMap[str, Any]
):
    image_name = Path(virtual_machine.image).name
    image_path = f"/tmp/{image_name}"

    files.download(
        name="Download cloud image",
        src=virtual_machine.image,
        dest=image_path
    )

    image_info = host.get_fact(ImageInfo, image_path=image_path, _ignore_errors=True)

    create_volume_as(
        name="Create image volume for base image",
        pool='images',
        volume_name=image_name,
        volume_size=image_info['actual-size'],
    )

    volume_upload(
        name="Upload image to volume",
        pool='images',
        volume_name=image_name,
        image_src=image_path,
    )
    
    domain_name = virtual_machine.name
    # name is reserved for pyinfra operation
    del(virtual_machine['name'])
    virt_install(
        name=f"Install {domain_name}",
        domain_name=domain_name,
        **virtual_machine
    )
```

So, this does obviously not work because the fact is gathered before any operation is executed and the image file is therefore not available. I through this is related to #1104 and added `_ignore_errors=True` when gathering facts but does not help because then python raises an exception at `volume_size=image_info['actual-size']` that NoneType is not subscriptable.

Using [Dynamic Execution](https://docs.pyinfra.com/en/3.x/examples/dynamic_execution_deploy.html) would require to put everything depending on `image_info` into the callback and would break idempotency AFAIU.

The only workaround I come up with would be to split the download step and the installation step into separate deploys and executing  them individually but that would make the whole process a lot less pleasant. I search through the the issues and looked at other deploys but they mostly use old API calls. #781 contained a promising addition with `host.reload_fact(...)` but I cannot find the function.

Would it be possible to implement such a behavior based on the API, running the two deploys in respective states? I'm a bit reluctant to this step because I've the feeling that I'm just missing a small detail to implement it using the current CLI.

**Comment by Fizzadar** [comment, marked as answer]
@an-ky you could use [nested operations](https://docs.pyinfra.com/en/3.x/using-operations.html#nested-operations) here which would be idempotent? Basically the same as dynamic execution but using operations.

Sidenote: let's update the dynamic execution doc to do this, the `run_shell_command` stuff is no longer needed most of the time (https://github.com/pyinfra-dev/pyinfra/issues/1241).

**Reply by an-ky** [reply]
Oh thanks, that solves the problem. It might a bit inconvenient at first because the actual flow of steps is broken up making reasoning about a bit harder but one gets used to it.

Still, I wonder how hard it would be to implement something like `host.dynamic_fact(...)` to query fact dynamically?
